<a href="https://colab.research.google.com/github/phurisk/BSC_DPDM23/blob/main/Frequent_Patterns_(Association_Rules).ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

In [7]:
import pandas as pd
import os

In [9]:
from google.colab import drive
drive.mount('/content/drive')
path = '/content/drive/MyDrive/BSC_DPDM23/File'

Drive already mounted at /content/drive; to attempt to forcibly remount, call drive.mount("/content/drive", force_remount=True).


In [10]:
data = pd.read_csv(os.path.join(path,'reduced_marketbasket.csv'),encoding='latin-1')

In [11]:
data

InvoiceNo StockCode                          Description  Quantity  \
0         536365    85123A   WHITE HANGING HEART T-LIGHT HOLDER         6   
1         536365     71053                  WHITE METAL LANTERN         6   
2         536365    84406B       CREAM CUPID HEARTS COAT HANGER         8   
3         536365    84029G  KNITTED UNION FLAG HOT WATER BOTTLE         6   
4         536365    84029E       RED WOOLLY HOTTIE WHITE HEART.         6   
...          ...       ...                                  ...       ...   
541904    581587     22613          PACK OF 20 SPACEBOY NAPKINS        12   
541905    581587     22899         CHILDREN'S APRON DOLLY GIRL          6   
541906    581587     23254        CHILDRENS CUTLERY DOLLY GIRL          4   
541907    581587     23255      CHILDRENS CUTLERY CIRCUS PARADE         4   
541908    581587     22138        BAKING SET 9 PIECE RETROSPOT          3   

             InvoiceDate  UnitPrice  CustomerID         Country  
0       12/01/2010 08:26       2.55     17850.0  United Kingdom  
1       12/01/2010 08:26       3.39     17850.0  United Kingdom  
2       12/01/2010 08:26       2.75     17850.0  United Kingdom  
3       12/01/2010 08:26       3.39     17850.0  United Kingdom  
4       12/01/2010 08:26       3.39     17850.0  United Kingdom  
...                  ...        ...         ...             ...  
541904  12/09/2011 12:50       0.85     12680.0          France  
541905  12/09/2011 12:50       2.10     12680.0          France  
541906  12/09/2011 12:50       4.15     12680.0          France  
541907  12/09/2011 12:50       4.15     12680.0          France  
541908  12/09/2011 12:50       4.95     12680.0          France  

[541909 rows x 8 columns]

## แปลง data ให้อยู่ในรูปแบบ transaction data ด้วย group by .appy(list)

In [13]:
data.groupby('InvoiceNo')['StockCode'].apply(list)

InvoiceNo
536365     [85123A, 71053, 84406B, 84029G, 84029E, 22752,...
536366                                        [22633, 22632]
536367     [84879, 22745, 22748, 22749, 22310, 84969, 226...
536368                          [22960, 22913, 22912, 22914]
536369                                               [21756]
                                 ...                        
C581484                                              [23843]
C581490                                       [22178, 23144]
C581499                                                  [M]
C581568                                              [21258]
C581569                                       [84978, 20979]
Name: StockCode, Length: 25900, dtype: object

In [14]:
transactions = data.groupby('InvoiceNo')['StockCode'].apply(list)

- ดูประเทศ

In [ ]:
set(data["Country"])

## Association Rules

In [15]:
!pip install apyori

  Preparing metadata (setup.py) ... done
  Created wheel for apyori: filename=apyori-1.1.2-py3-none-any.whl size=5954 sha256=8102d12945cd559040683e3213a433e0d3d56dc6649ef85f42fe245def43f057
  Stored in directory: /root/.cache/pip/wheels/c4/1a/79/20f55c470a50bb3702a8cb7c94d8ada15573538c7f4baebe2d
Successfully built apyori


In [16]:
from apyori import apriori

In [17]:
frequentItemsets = list(apriori(transactions,min_support = 0.065))
frequentItemsets

[RelationRecord(items=frozenset({'22423'}), support=0.08386100386100385, ordered_statistics=[OrderedStatistic(items_base=frozenset(), items_add=frozenset({'22423'}), confidence=0.08386100386100385, lift=1.0)]),
 RelationRecord(items=frozenset({'47566'}), support=0.06586872586872587, ordered_statistics=[OrderedStatistic(items_base=frozenset(), items_add=frozenset({'47566'}), confidence=0.06586872586872587, lift=1.0)]),
 RelationRecord(items=frozenset({'85099B'}), support=0.08243243243243244, ordered_statistics=[OrderedStatistic(items_base=frozenset(), items_add=frozenset({'85099B'}), confidence=0.08243243243243244, lift=1.0)]),
 RelationRecord(items=frozenset({'85123A'}), support=0.08671814671814672, ordered_statistics=[OrderedStatistic(items_base=frozenset(), items_add=frozenset({'85123A'}), confidence=0.08671814671814672, lift=1.0)])]

In [18]:
len(frequentItemsets)

4

In [19]:
frequentItemsets[0][0]

frozenset({'22423'})

In [20]:
dir(frequentItemsets[0][0])

['__and__',
 '__class__',
 '__class_getitem__',
 '__contains__',
 '__delattr__',
 '__dir__',
 '__doc__',
 '__eq__',
 '__format__',
 '__ge__',
 '__getattribute__',
 '__gt__',
 '__hash__',
 '__init__',
 '__init_subclass__',
 '__iter__',
 '__le__',
 '__len__',
 '__lt__',
 '__ne__',
 '__new__',
 '__or__',
 '__rand__',
 '__reduce__',
 '__reduce_ex__',
 '__repr__',
 '__ror__',
 '__rsub__',
 '__rxor__',
 '__setattr__',
 '__sizeof__',
 '__str__',
 '__sub__',
 '__subclasshook__',
 '__xor__',
 'copy',
 'difference',
 'intersection',
 'isdisjoint',
 'issubset',
 'issuperset',
 'symmetric_difference',
 'union']

In [21]:
frequentItemsets[0][0].__len__()

1

In [34]:
min_sup = 0.03

frequentItemsets = list(apriori(transactions,min_support = min_sup))
for i in frequentItemsets:
    if i[0].__len__()>1:
        print('yes')
        print(i)

yes
RelationRecord(items=frozenset({'22386', '85099B'}), support=0.03216216216216216, ordered_statistics=[OrderedStatistic(items_base=frozenset(), items_add=frozenset({'22386', '85099B'}), confidence=0.03216216216216216, lift=1.0), OrderedStatistic(items_base=frozenset({'22386'}), items_add=frozenset({'85099B'}), confidence=0.6766856214459789, lift=8.20897311262335), OrderedStatistic(items_base=frozenset({'85099B'}), items_add=frozenset({'22386'}), confidence=0.39016393442622943, lift=8.20897311262335)])
yes
RelationRecord(items=frozenset({'22697', '22699'}), support=0.03027027027027027, ordered_statistics=[OrderedStatistic(items_base=frozenset(), items_add=frozenset({'22697', '22699'}), confidence=0.03027027027027027, lift=1.0), OrderedStatistic(items_base=frozenset({'22697'}), items_add=frozenset({'22699'}), confidence=0.7417218543046358, lift=17.1523178807947), OrderedStatistic(items_base=frozenset({'22699'}), items_add=frozenset({'22697'}), confidence=0.7, lift=17.1523178807947)])


RelationRecord(items=frozenset({'22386', '85099B'}),

support=0.03216216216216216,

ordered_statistics=[OrderedStatistic(items_base=frozenset(), items_add=frozenset({'22386', '85099B'}), confidence=0.03216216216216216, lift=1.0),

OrderedStatistic(items_base=frozenset({'22386'}), items_add=frozenset({'85099B'}), confidence=0.6766856214459789, lift=8.20897311262335),

OrderedStatistic(items_base=frozenset({'85099B'}), items_add=frozenset({'22386'}), confidence=0.39016393442622943, lift=8.20897311262335)])


In [35]:
set(data[(data['StockCode']=='22386') | (data['StockCode']=='85099B') | (data['StockCode']=='22699') | (data['StockCode']=='22697') ]['Description'])

{'GREEN REGENCY TEACUP AND SAUCER',
 'JUMBO BAG PINK POLKADOT',
 'JUMBO BAG RED RETROSPOT',
 'ROSES REGENCY TEACUP AND SAUCER '}

RelationRecord(items=frozenset({'22699', '22697'}), support=0.03027027027027027,

ordered_statistics=[OrderedStatistic(items_base=frozenset(), items_add=frozenset({'22699', '22697'}), confidence=0.03027027027027027, lift=1.0),

OrderedStatistic(items_base=frozenset({'22697'}), items_add=frozenset({'22699'}), confidence=0.7417218543046358, lift=17.1523178807947),

OrderedStatistic(items_base=frozenset({'22699'}), items_add=frozenset({'22697'}), confidence=0.7, lift=17.1523178807947)])

In [ ]:
set(data[(data['StockCode']=='22699') | (data['StockCode']=='22697')]['InvoiceNo'])

In [ ]:
data[data['InvoiceNo']=='562950']

## quiz 22386 และ 85099b คือสินค้าอะไร

In [47]:
set(data[(data['StockCode']=='22386')  ]['Description'])

{'JUMBO BAG PINK POLKADOT'}

In [46]:
set(data[(data['StockCode']=='85099B') ]['Description'])

{'JUMBO BAG RED RETROSPOT'}

#Hw หา Frequent Iten Set  และ Association Rules
- ของช่วงเวลา Chirtmas & New Year(1 Dec -31 Dec)
- ของประเทศในทวีป Europe
- ของช่วงเวลา ตรุษจีน ในทวีป Asai (15 jan-15 Feb)
- ของช่วงเวลา Chirtmas & New Year ในทวีป Europe (1 Dec -31 Dec)
- ของวันวาเลนไทน์
- โดยจัดกลุ่มตามประเทศ+เดือน

In [51]:
# Import necessary libraries
import pandas as pd
from mlxtend.frequent_patterns import apriori, association_rules
from mlxtend.preprocessing import TransactionEncoder
from datetime import datetime

data['InvoiceDate'] = pd.to_datetime(data['InvoiceDate'])

# Define the time periods and regions
time_periods = {
    "Christmas_NewYear_Europe": {
        "start": datetime(2010, 12, 1),
        "end": datetime(2010, 12, 31),
        "regions": ['List', 'Of', 'European', 'Countries']  # Replace with actual country names
    },
    "Chinese_New_Year_Asia": {
        "start": datetime(2011, 1, 15),
        "end": datetime(2011, 2, 15),
        "regions": ['List', 'Of', 'Asian', 'Countries']  # Replace with actual country names
    },
    "Valentines_Day": {
        "start": datetime(2011, 2, 14),
        "end": datetime(2011, 2, 14),
        "regions": None  # All regions
    }
}

# Function to preprocess data for each period
def preprocess_data(data, start_date, end_date, countries=None):
    period_data = data[(data['InvoiceDate'] >= start_date) & (data['InvoiceDate'] <= end_date)]
    if countries:
        period_data = period_data[period_data['Country'].isin(countries)]
    period_data['Month'] = period_data['InvoiceDate'].dt.month
    return period_data

# Function to perform analysis for each period
def perform_analysis(data):
    results = {}
    for country, group_data in data.groupby('Country'):
        for month, month_data in group_data.groupby('Month'):
            transactions = month_data.groupby('InvoiceNo')['Description'].apply(list)
            te = TransactionEncoder()
            te_ary = te.fit(transactions).transform(transactions)
            df = pd.DataFrame(te_ary, columns=te.columns_)
            frequent_item_sets = apriori(df, min_support=0.01, use_colnames=True)
            rules = association_rules(frequent_item_sets, metric="confidence", min_threshold=0.1)
            results[(country, month)] = rules
    return results

# Function to find frequent item sets and association rules for a specific period and region
def find_association_rules(period_name, start_date, end_date, countries):
    period_data = preprocess_data(data, start_date, end_date, countries)
    analysis_results = perform_analysis(period_data)

    for key, rules in analysis_results.items():
        print(f"Period: {period_name}")
        print(f"Country: {key[0]}, Month: {key[1]}")
        print(rules)

# Find Frequent Item Sets and Association Rules

# Christmas & New Year period (1 Dec - 31 Dec) in Europe
find_association_rules("Christmas_NewYear_Europe", datetime(2010, 12, 1), datetime(2010, 12, 31), ['List', 'Of', 'European', 'Countries'])

# Chinese New Year period in Asia (15 Jan - 15 Feb)
find_association_rules("Chinese_New_Year_Asia", datetime(2011, 1, 15), datetime(2011, 2, 15), ['List', 'Of', 'Asian', 'Countries'])

# Christmas & New Year period in Europe (1 Dec - 31 Dec)
find_association_rules("Christmas_NewYear_Europe", datetime(2010, 12, 1), datetime(2010, 12, 31), None)  # All regions

# Valentine's Day
find_association_rules("Valentines_Day", datetime(2011, 2, 14), datetime(2011, 2, 14), None)  # All regions


/usr/local/lib/python3.10/dist-packages/ipykernel/ipkernel.py:283: DeprecationWarning: `should_run_async` will not call `transform_cell` automatically in the future. Please pass the result to `transformed_cell` argument and any exception that happen during thetransform in `preprocessing_exc_tuple` in IPython 7.17 and above.
  and should_run_async(code)
<ipython-input-51-20f89e4377dc>:33: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  period_data['Month'] = period_data['InvoiceDate'].dt.month


KeyboardInterrupt: ignored